[WPE] Implement web view maximize/minimize/restore window when using the new API

siwei-li opened the pull request:
#### 0cfd8f953c128c09d6240b68b9f89c2330e7c747
<pre>
[WPE] Implement web view maximize/minimize/restore window when using the new API
<a href="https://bugs.webkit.org/show_bug.cgi?id=277945">https://bugs.webkit.org/show_bug.cgi?id=277945</a>

Reviewed by NOBODY (OOPS!).

Add toplevel minimization/maximization/restoration support to the webdriver.
Wayland compositor does not notify when the minimization actually happens,
hence we only ask the toplevel to be minimized.

* Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
(ToplevelStateEvent::ToplevelStateEvent):
(ToplevelStateEvent::~ToplevelStateEvent):
(ToplevelStateEvent::complete):
(finishMonitorToplevelState):
(toplevelStateChangedCallback):
(monitorToplevelState):
(webkitWebViewMaximizeWindow):
(webkitWebViewMinimizeWindow):
(webkitWebViewRestoreWindow):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0cfd8f953c128c09d6240b68b9f89c2330e7c747

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63335 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42691 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15932 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67356 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13943 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50379 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14223 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51020 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9637 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66404 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39641 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54843 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31701 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36323 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12815 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57863 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12520 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69052 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7282 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12129 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58333 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7313 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54913 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58597 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6068 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38512 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39591 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40703 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39334 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,136 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
+}
+
+static void
+monitorToplevelState(WPEView* view, ToplevelStateEvent::Type type, CompletionHandler<void()>&& completionHandler)
```
```suggestion
static void monitorToplevelState(WPEView* view, ToplevelStateEvent::Type type, CompletionHandler<void()>&& completionHandler)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,136 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
+
+        monitorToplevelState(wpeView, ToplevelStateEvent::Type::Maximize, WTFMove(completionHandler));
+        if (wpe_toplevel_maximize(toplevel))
+            return;
```
The problem here is that if maximize is not supported we don't return here, but completionHandler was already moved and it's now handled by ToplevelState. So, I would add another helper `finishMonitorToplevelState()` that disconnects the function and sets the user data to nullptr, to call it from here before returning and also from toplevelStateChangedCallback when event was completed.

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,136 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
+
+        monitorToplevelState(wpeView, ToplevelStateEvent::Type::Minimize, WTFMove(completionHandler));
+        if (wpe_toplevel_minimize(toplevel))
+            return;
```
Same happens here.

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,136 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
+
+        monitorToplevelState(wpeView, ToplevelStateEvent::Type::Restore, WTFMove(completionHandler));
+        if ((wpe_toplevel_get_state(toplevel) & WPE_TOPLEVEL_STATE_MAXIMIZED) && wpe_toplevel_unmaximize(toplevel))
+            return;
```
And here.

siwei-li committed:
[WPE] Implement web view maximize/minimize/restore window when using the new API
https://bugs.webkit.org/show_bug.cgi?id=277945

Reviewed by NOBODY (OOPS!).

Add toplevel minimization/maximization/restoration support to the webdriver.
Wayland compositor does not notify when the minimization actually happens,
hence we only ask the toplevel to be minimized.

* Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
(ToplevelStateEvent::ToplevelStateEvent):
(ToplevelStateEvent::~ToplevelStateEvent):
(ToplevelStateEvent::complete):
(finishMonitorToplevelState):
(toplevelStateChangedCallback):
(monitorToplevelState):
(webkitWebViewMaximizeWindow):
(webkitWebViewMinimizeWindow):
(webkitWebViewRestoreWindow):
